### PR TITLE
Python 3

### PR DESCRIPTION
**Pull request by ro6ley**
#### What does this PR do?
Port Indigo to Python 3

#### Description of Task to be completed?
Python 2 support is running out on January 1, 2020. This PR ports the platform to the latest Python 3.7.

#### How should this be manually tested?
- Set up Indigo as per the instructions on the README.
- Run `python manage.py test`

#### Known issues 
- `Pinax Badges` does not work on Python 3.7
- Latest `Cobalt` not yet on PyPi, currently installing from <a href="https://github.com/laws-africa/cobalt/tree/python3" target="_blank">github</a>
- Currently works on Python 3.6 and not Python 3.7 due to the Pinax-Badges dependency

### CODE REVIEW

File: indigo/analysis/terms/base.py [hunk 1]
```diff
@@ -176,10 +174,10 @@ def find_term_references(self, doc, terms):
             return
 
         # term to term id
-        term_lookup = {v: k for k, v in terms.iteritems()}
+        term_lookup = {v: k for k, v in terms.items()}
 
         # big regex of all the terms, longest first
-        terms = sorted(terms.itervalues(), key=lambda t: -len(t))
+        terms = sorted(iter(terms.values()), key=lambda t: -len(t))
```
Comment (longhotsummer) [inline indigo/analysis/terms/base.py]: I don't think `iter` is necessary.

```suggestion
        terms = sorted(terms.values(), key=lambda t: -len(t))
```

File: indigo/bulk_creator.py [hunk 2]
```diff
@@ -123,9 +120,8 @@ def get_datatable_gsheets(self, spreadsheet_id, sheet_name):
         index (tab index).
         """
         try:
-            result = self.gsheets_client\
-                .spreadsheets()\
-                .values().get(spreadsheetId=spreadsheet_id, range=sheet_name)\
+            result = list(self.gsheets_client\
```
Comment (longhotsummer) [inline indigo/bulk_creator.py]: Was this an automatic correction? I don't think it's correct. The call to `values()` isn't a dictionary call. I think the whole thing can remain unchanged.

File: indigo/bulk_creator.py [hunk 3]
```diff
@@ -167,7 +163,7 @@ def create_works(self, view, table, dry_run, workflow, user):
 
         for idx, row in enumerate(rows):
             # ignore if it's blank or explicitly marked 'ignore' in the 'ignore' column
-            if row.get('ignore') or not [val for val in row.values() if val]:
+            if row.get('ignore') or not [val for val in list(row.values()) if val]:
```
Comment (longhotsummer) [inline indigo/bulk_creator.py]: No need for `list()`, we're just iterating over the values.
```suggestion
            if row.get('ignore') or not [val for val in row.values() if val]:
```

File: indigo/bulk_creator.py [hunk 4]
```diff
@@ -259,10 +255,10 @@ def create_work(self, view, row, idx, dry_run):
                 info['status'] = 'error'
                 if hasattr(e, 'message_dict'):
                     info['error_message'] = ' '.join(
-                        ['%s: %s' % (f, '; '.join(errs)) for f, errs in e.message_dict.items()]
+                        ['%s: %s' % (f, '; '.join(errs)) for f, errs in list(e.message_dict.items())]
```
Comment (longhotsummer) [inline indigo/bulk_creator.py]:
```suggestion
                        ['%s: %s' % (f, '; '.join(errs)) for f, errs in e.message_dict.items()]
```

File: indigo/bulk_creator.py [hunk 5]
```diff
@@ -325,7 +321,7 @@ def get_frbr_uri(self, row):
         return frbr_uri.work_uri().lower()
 
     def add_extra_properties(self, work, info):
-        for extra_property in self.extra_properties.keys():
+        for extra_property in list(self.extra_properties.keys()):
```
Comment (longhotsummer) [inline indigo/bulk_creator.py]:
```suggestion
        for extra_property in self.extra_properties.keys():
```

File: indigo/plugins.py [hunk 6]
```diff
@@ -30,7 +30,7 @@ def for_locale(self, topic, country=None, language=None, locality=None):
         """ Find an appropriate importer for this locale description. Tightest match wins.
         """
         target = (country, language, locality)
-        match = self.lookup(topic, target, self.registry[topic].itervalues())
+        match = self.lookup(topic, target, iter(self.registry[topic].values()))
```
Comment (longhotsummer) [inline indigo/plugins.py]: `.values()` is already an iterable, doesn't need `iter()`
```suggestion
        match = self.lookup(topic, target, self.registry[topic].values())
```

File: indigo/strftime_patch.py [hunk 7]
```diff
@@ -79,7 +79,7 @@ def test():
     if s != "1800 has the same days as 1980 and 2008":
```
Comment (longhotsummer) [inline indigo/strftime_patch.py]: The entire strftime_patch can be removed for python3.

File: indigo_api/migrations/0037_document_work.py [hunk 8]
```diff
@@ -27,7 +25,7 @@ def create_works(apps, schema_editor):
             commencement_date=d.commencement_date,
             assent_date=d.assent_date,
             deleted=d.deleted,
-        ) for uri, d in work_detail.iteritems()
+        ) for uri, d in list(work_detail.items())
```
Comment (longhotsummer) [inline indigo_api/migrations/0037_document_work.py]:
```suggestion
        ) for uri, d in work_detail.items()
```

File: indigo_api/models.py [hunk 9]
```diff
@@ -51,7 +51,7 @@ def code(self):
         return self.language.iso_639_2B
 
     def __unicode__(self):
```
Comment (longhotsummer) [inline indigo_api/models.py]: In python3, `__unicode__` should be `__str__`

File: indigo_api/models.py [hunk 10]
```diff
@@ -330,7 +330,7 @@ def labeled_properties(self):
             'label': WorkProperty.KEYS[key],
             'key': key,
             'value': val,
-        } for key, val in self.properties.iteritems() if key in WorkProperty.KEYS], key=lambda x: x['label'])
+        } for key, val in list(self.properties.items()) if key in WorkProperty.KEYS], key=lambda x: x['label'])
```
Comment (longhotsummer) [inline indigo_api/models.py]:
```suggestion
        } for key, val in self.properties.items() if key in WorkProperty.KEYS], key=lambda x: x['label'])
```

File: indigo_api/models.py [hunk 11]
```diff
@@ -1423,7 +1423,7 @@ def grouper(task):
                 'badge': key,
             }
 
-        for key, group in tasks.iteritems():
+        for key, group in list(tasks.items()):
```
Comment (longhotsummer) [inline indigo_api/models.py]:
```suggestion
        for key, group in tasks.items():
```

File: indigo_api/renderers.py [hunk 12]
```diff
@@ -86,13 +86,13 @@ def render(self, node):
         params = {
             'defaultIdScope': ET.XSLT.strparam(self.defaultIdScope(node) or ''),
         }
-        params.update({k: ET.XSLT.strparam(v) for k, v in self.xslt_params.iteritems()})
+        params.update({k: ET.XSLT.strparam(v) for k, v in list(self.xslt_params.items())})
```
Comment (longhotsummer) [inline indigo_api/renderers.py]: No `list`

File: indigo_api/serializers.py [hunk 13]
```diff
@@ -389,7 +389,7 @@ def update_document(self, document, validated_data=None):
             document.content = content
 
         # save rest of changes
-        for attr, value in validated_data.items():
+        for attr, value in list(validated_data.items()):
```
Comment (longhotsummer) [inline indigo_api/serializers.py]: No `list`

File: indigo_app/forms.py [hunk 14]
```diff
@@ -1,5 +1,5 @@
 import json
-import urllib
+import urllib.request, urllib.parse, urllib.error
```
Comment (longhotsummer) [inline indigo_app/forms.py]:
```suggestion
import urllib.parse
```

File: indigo_content_api/v2/views.py [hunk 15]
```diff
@@ -32,7 +30,7 @@ def rewrite_frbr_uris(self, data):
         start with /akn/
         """
         if isinstance(data, dict):
-            for key, val in data.iteritems():
+            for key, val in list(data.items()):
```
Comment (longhotsummer) [inline indigo_content_api/v2/views.py]: No `list`

File: indigo_social/views.py [hunk 16]
```diff
@@ -256,7 +254,7 @@ class BadgeListView(TemplateView):
     template_name = 'indigo_social/badges.html'
 
     def get_context_data(self, **context):
-        context['badges'] = sorted(badges.registry.values(), key=lambda b: b.name)
+        context['badges'] = sorted(list(badges.registry.values()), key=lambda b: b.name)
```
Comment (longhotsummer) [inline indigo_social/views.py]: No `list`

File: indigo/bulk_creator.py (hunk 2 above)
Comment (ro6ley) [reply indigo/bulk_creator.py]: Yes, this was an automatic one, as were the other `list()` changes

File: indigo_resolver/authorities.py [hunk 17]
```diff
@@ -20,7 +20,7 @@ def get(self, name):
         return Authority.objects.filter(slug=name).first()
 
     def all(self):
-        return list(self.registry.values()) + list(Authority.objects.all())
+        return self.registry.values() + list(Authority.objects.all())
```
Comment (longhotsummer) [inline indigo_resolver/authorities.py]: Needs list

File: indigo_app/views/places.py [hunk 18]
```diff
@@ -368,7 +368,7 @@ def subtype_name(abbr):
                 return 'Act'
             st = Subtype.for_abbreviation(abbr)
             return st.name if st else abbr
-        pairs = list(Counter([subtype_name(w.subtype) for w in works]).items())
+        pairs = Counter([subtype_name(w.subtype) for w in works]).items()
```
Comment (longhotsummer) [inline indigo_app/views/places.py]: Needs list

File: indigo_app/views/places.py [hunk 19]
```diff
@@ -358,7 +358,7 @@ def get_context_data(self, **kwargs):
             .all()
         years = {x['year']: x['n'] for x in years}
         self.add_year_zeros(years)
-        years = list(years.items())
+        years = years.items()
```
Comment (longhotsummer) [inline indigo_app/views/places.py]: Needs list

File: indigo_app/views/places.py [hunk 20]
```diff
@@ -344,7 +344,7 @@ def get_context_data(self, **kwargs):
             .select_related(None).prefetch_related(None).all()
         years = Counter([int(w.year) for w in works])
         self.add_year_zeros(years)
-        years = list(years.items())
+        years = years.items()
```
Comment (longhotsummer) [inline indigo_app/views/places.py]: Needs list

File: indigo_app/forms.py [hunk 21]
```diff
@@ -106,13 +106,13 @@ class BaseWorkPropertyFormSet(BaseModelFormSet):
...
         existing = set([p.key for p in self.queryset.all()])
-        missing = [key for key in list(WorkProperty.KEYS.keys()) if key not in existing]
+        missing = [key for key in WorkProperty.KEYS.keys() if key not in existing]
         self.extra = len(missing)
         self.initial_extra = [{'key': key} for key in missing]
 
     def keys_and_forms(self):
         # (value, label) pairs sorted by label
-        keys = sorted(list(WorkProperty.KEYS.items()), key=lambda x: x[1])
+        keys = sorted(WorkProperty.KEYS.items(), key=lambda x: x[1])
```
Comment (longhotsummer) [inline indigo_app/forms.py]: Fine

File: indigo_app/forms.py [hunk 22]
```diff
@@ -106,13 +106,13 @@ class BaseWorkPropertyFormSet(BaseModelFormSet):
     def setup_extras(self):
         # add extra forms for the properties we don't have yet
         existing = set([p.key for p in self.queryset.all()])
-        missing = [key for key in list(WorkProperty.KEYS.keys()) if key not in existing]
+        missing = [key for key in WorkProperty.KEYS.keys() if key not in existing]
```
Comment (longhotsummer) [inline indigo_app/forms.py]: Fine

File: indigo_app/forms.py [hunk 23]
```diff
@@ -93,7 +93,7 @@ class Meta:
 
     def __init__(self, *args, **kwargs):
         super(WorkPropertyForm, self).__init__(*args, **kwargs)
-        self.fields['key'].choices = list(WorkProperty.KEYS.items())
+        self.fields['key'].choices = WorkProperty.KEYS.items()
```
Comment (longhotsummer) [inline indigo_app/forms.py]: Needs list

Comment (ro6ley) [reply indigo_app/forms.py]: Added it

**Comment by longhotsummer** [comment]
Looks good, thanks Robley.

**Comment by longhotsummer** [comment]
Next step is to upgrade the other indigo packages that depend on this one, then we can merge them all to master.

**Pull request merged**